service/history/workflow/mutable_state_impl.go

```diff
@@ -5137,14 +5133,112 @@ func (ms *MutableStateImpl) prepareCloseTransaction(
 		}
 	}
 
+	return nil
+}
+
+func (ms *MutableStateImpl) closeTransactionHandleSpeculativeWorkflowTask(
+	transactionPolicy TransactionPolicy,
+) error {
+	if transactionPolicy == TransactionPolicyPassive ||
+		!ms.IsWorkflowExecutionRunning() {
+		return nil
+	}
+
+	// It is important to convert speculative WT to normal before prepareEventsAndReplicationTasks,
+	// because prepareEventsAndReplicationTasks will move internal buffered events to the history,
+	// and WT related events (WTScheduled, in particular) need to go first.
+	return ms.workflowTaskManager.convertSpeculativeWorkflowTaskToNormal()
+}
+
+func (ms *MutableStateImpl) closeTransactionUpdateTransitionHistory(
+	transactionPolicy TransactionPolicy,
+	workflowEventsSeq []*persistence.WorkflowEvents,
+) error {
+
+	if len(workflowEventsSeq) > 0 {
+		lastEvents := workflowEventsSeq[len(workflowEventsSeq)-1].Events
+		lastEvent := lastEvents[len(lastEvents)-1]
+		if err := ms.updateWithLastWriteEvent(
+			lastEvent,
+			transactionPolicy,
+		); err != nil {
+			return err
+		}
+	}
+
+	if transactionPolicy != TransactionPolicyActive {
```

Comment:
does it manke to check the policy before update the last write event?

Reply:
we don't have that check for last write event before as well. I think replication (passive) logic relies on that to update version history as well?